includes/gateway/class-omise-payment-billpayment-tesco.php

```diff
@@ -0,0 +1,226 @@
+					'type'    => 'checkbox',
+					'label'   => __( 'Enable Omise Tesco Bill Payment', 'omise' ),
+					'default' => 'no'
+				),
+
+				'title' => array(
+					'title'       => __( 'Title', 'omise' ),
+					'type'        => 'text',
+					'description' => __( 'This controls the title which the user sees during checkout.', 'omise' ),
+					'default'     => __( 'Bill Payment: Tesco', 'omise' ),
+				),
+
+				'description' => array(
+					'title'       => __( 'Description', 'omise' ),
+					'type'        => 'textarea',
+					'description' => __( 'This controls the description which the user sees during checkout.', 'omise' )
+				),
+			);
+		}
+
+		/**
+		 * @inheritdoc
+		 */
+		public function charge( $order_id, $order ) {
+			$metadata = array_merge(
+				apply_filters( 'omise_charge_params_metadata', array(), $order ),
+				array( 'order_id' => $order_id ) // override order_id as a reference for webhook handlers.
+			);
+
+			return OmiseCharge::create( array(
+				'amount'      => Omise_Money::to_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters( 'omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order ),
+				'source'      => array( 'type' => 'bill_payment_tesco_lotus' ),
+				'return_uri'  => add_query_arg( 'order_id', $order_id, site_url() . "?wc-api=omise_billpayment_tesco_callback" ),
+				'metadata'    => $metadata
+			) );
+		}
+
+		/**
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( 'failed' == $charge['status'] ) {
+				return $this->payment_failed( $charge['failure_message'] . ' (code: ' . $charge['failure_code'] . ')' );
+			}
+
+			if ( 'pending' == $charge['status'] ) {
+				return [
+					'result'   => 'success',
+					'redirect' => $this->get_return_url( $order )
+				];
+			}
+
+			return $this->payment_failed(
+				sprintf(
+					__( 'Please feel free to try submitting your order again, or contact our support team if you have any questions (Your temporary order id is \'%s\')', 'omise' ),
+					$order_id
+				)
+			);
+		}
+
+		/**
+		 * @param int|WC_Order $order
+		 * @param string       $context  pass 'email' value through this argument only for 'sending out an email' case.
+		 */
+		public function display_barcode( $order, $context = 'view' ) {
+			if ( ! $this->load_order( $order ) ) {
+				return;
+			}
+
+			$charge_id    = $this->get_charge_id_from_order();
+			$charge       = OmiseCharge::retrieve( $charge_id );
+			$barcode_svg  = file_get_contents( $charge['source']['references']['barcode'] );
+			?>
+
+			<div class="omise omise-billpayment-tesco-details" <?php echo 'email' === $context ? 'style="margin-bottom: 4em; text-align:center;"' : ''; ?>>
+				<p><?php echo __( 'Use this barcode to pay at Tesco Lotus.', 'omise' ); ?></p>
+				<div class="omise-billpayment-tesco-barcode">
+					<?php
+					if ( 'email' === $context ) {
```

Comment:
Do the emails have plaintext version too? Just wondering what happens if the email was opened on a non-html client (however rare that might be). Do we include a link to the barcode?

Reply:
@jonrandy yup, there is a plaintext email version too in WC email setting.
Can you name me some non-html client that we should test? no idea on that.

Reply:
I know Mozilla Thunderbird can be made to open emails in plain text by default

Reply:
Small article on plain text emails - https://litmus.com/blog/best-practices-for-plain-text-emails-a-look-at-why-theyre-important

Reply:
@jonrandy Come to think of it, what do you think if we put this topic aside for now as it's not serve a majority users?

Supporting plaintext email requires a lot more stuff to do (i.e. generate a new page in WordPress system so we can put the generated page's link to an order confirmation email) and it might not worth to priority it at the moment.